pkg/controller/bastion/options.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+
+	openstackapi "github.com/gardener/gardener-extension-provider-openstack/pkg/apis/openstack"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/rules"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	//maxLengthForBaseName for "base" name due to fact that we use this name to name other openstack resources,
+	maxLengthForBaseName = 33
+)
+
+// Options contains provider-related information required for setting up
+// a bastion instance. This struct combines precomputed values like the
+// bastion instance name with the IDs of pre-existing cloud provider
+// resources, like the nic name, subnet name etc.
+type Options struct {
+	BastionInstanceName string
+	Region              string
+	ShootName           string
+	SecretReference     corev1.SecretReference
+	SecurityGroup       string
+	UserData            []byte
+	FloatingPoolName    string
+}
+
+// DetermineOptions determines the required information that are required to reconcile a Bastion on Openstack. This
+// function does not create any IaaS resources.
+func DetermineOptions(bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) (*Options, error) {
+	clusterName := cluster.ObjectMeta.Name
+	region := cluster.Shoot.Spec.Region
+
+	baseResourceName, err := generateBastionBaseResourceName(clusterName, bastion.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	secretReference := corev1.SecretReference{
+		Namespace: clusterName,
+		Name:      v1beta1constants.SecretNameCloudProvider,
+	}
+
+	infrastructureConfig := &openstackapi.InfrastructureConfig{}
+	err = json.Unmarshal(cluster.Shoot.Spec.Provider.InfrastructureConfig.Raw, infrastructureConfig)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
If I see correctly, in all places we use `decoder.Decode` to decode a RawExtension. 

https://github.com/gardener/gardener-extension-provider-openstack/blob/0deb7a8c89d100dec809c75e915404231876e02d/pkg/admission/validator/serialization.go#L33-L40

Reply:
thanks let me know. 7b1e67a9